### PR TITLE
Populate `process.command_args` for Java 9+ and improve `process.command_line` for Java 8

### PR DESCRIPTION
**Pull request by tylerbenson**
`command_args` is defined as:
> All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from proc/[pid]/cmdline. For libc-based executables, this would be the full argv vector passed to main.

`process.command_line` only included vm arguments, not the program arguments:
> The full command used to launch the process as a single string representing the full command. On Windows, can be set to the result of GetCommandLineW. Do not set this if you have to assemble it just for monitoring; use process.command_args instead.

Based on review below, I opted to only provide `command_line` for Java 8 and prefer `command_args` for Java 9+.

### CODE REVIEW

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java [hunk 1]
```diff
@@ -69,10 +72,33 @@ private static Resource doBuildResource() {
 
       attributes.put(ResourceAttributes.PROCESS_EXECUTABLE_PATH, executablePath.toString());
 
+      List<String> commandArgs = new ArrayList<>();
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java:75]: Consider using https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ProcessHandle.Info.html With that you should be able to get what is needed on jdk9. If we have a good implementation for jdk9+ then perhaps it isn't too important that our jdk8 implementation isn't as good.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java [hunk 2]
```diff
@@ -69,10 +72,33 @@ private static Resource doBuildResource() {
...
+      // sun.java.command doesn't seem to be well document or official, but there
+      // doesn't seem to be any better way to get this consistently without more complexity.
+      String javaCommand = System.getProperty("sun.java.command");
+      if (javaCommand != null) {
+        String[] args = javaCommand.split(" ");
+        if (args.length > 0) {
+          // TODO: add handling for windows paths.
+          if (args[0].startsWith("/")) {
+            // We are dealing with a `java -jar /path/to/some.jar` situation and need to add
+            commandArgs.add("-jar");
```
Comment (laurit) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java:87]: Should prepend `-jar ` to `javaCommand ` to have it in `process.command_line` too?

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java:75]: Alternatively, other resource providers use MR jar: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/resources/library/src/main/java9/io/opentelemetry/instrumentation/resources/ProcessArguments.java

Comment (tylerbenson) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java:75]: I've added @mateuszrzeszutek suggestion when it's available. Still falls back to the current implementation though.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java [hunk 3]
```diff
@@ -69,10 +72,33 @@ private static Resource doBuildResource() {
...
+      commandArgs.add(executablePath.toString());
+      commandArgs.addAll(runtime.getInputArguments()); // Only includes VM arguments
+      // sun.java.command doesn't seem to be well document or official, but there
+      // doesn't seem to be any better way to get this consistently without more complexity.
+      String javaCommand = System.getProperty("sun.java.command");
+      if (javaCommand != null) {
+        String[] args = javaCommand.split(" ");
+        if (args.length > 0) {
+          // TODO: add handling for windows paths.
+          if (args[0].startsWith("/")) {
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java:86]: There's no requirement for executable jar files to end with .jar.  For example, I've seen examples of executable war files.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java [hunk 4]
```diff
@@ -69,10 +72,33 @@ private static Resource doBuildResource() {
...
       attributes.put(ResourceAttributes.PROCESS_EXECUTABLE_PATH, executablePath.toString());
 
+      List<String> commandArgs = new ArrayList<>();
+      commandArgs.add(executablePath.toString());
+      commandArgs.addAll(runtime.getInputArguments()); // Only includes VM arguments
+      // sun.java.command doesn't seem to be well document or official, but there
+      // doesn't seem to be any better way to get this consistently without more complexity.
+      String javaCommand = System.getProperty("sun.java.command");
+      if (javaCommand != null) {
+        String[] args = javaCommand.split(" ");
```
[the remark this answers is not in the document]
Comment (tylerbenson) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java]: Ok, I set one or the other, but not both depending on if the individual args are available.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessArguments.java [hunk 5]
```diff
@@ -5,8 +5,11 @@
 
 package io.opentelemetry.instrumentation.resources;
 
+import javax.annotation.Nonnull;
+
 final class ProcessArguments {
 
+  @Nonnull
```
Comment (mateuszrzeszutek) [inline instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessArguments.java:12]: Nit: we only use `@Nullable`, everything not annotated with anything should be non-null by default.

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java]: I think the important thing here is to recognize that we can't accurately tell whether user ran `java -jar somejar` or `java someclass`. Which gives use 2 choices, either we don't add `-jar` at all or guess. I'd replace the `// TODO: add handling for windows paths.` with a comment that indicates that we are trying to guess whether the argument is a jar or not. `args[0].startsWith("/")` only works when you give absolute path to jar while `args[0].endsWith(".jar")` would also work with relative paths, but wouldn't work with executable war and could give a false positive when someone is executing a class named `jar`. Could also use `args[0].startsWith("/") || args[0].endsWith(".jar")`. Could also try something like `ClassLoader.getSystemClassLoader().getResource(args[0].replace('.', '/') + ".class") != null` -> class resource exists so probably not a jar, but I don't think this is worth the effort.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessArguments.java (hunk 5 above)
Comment (tylerbenson) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessArguments.java]: Ok, removed.

File: instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tylerbenson) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java]: I've switched to using a regex instead.  LMK if this is sufficient.  (Still won't handle if the file path has a space in it though.)

[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources/ProcessResource.java]: I think it is fine

**Review by trask — APPROVED** [review]

**Pull request merged**